sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -417,15 +422,25 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
       // needs a particular sort order, it should not be a problem in practice that would
       // trigger a redundant sort in the plan.
       plan match {
-        case _: GpuHashJoin =>
-          val gpuSortOrder = getOptimizedSortOrder(plan)
-          GpuSortExec(gpuSortOrder, false, plan, SortEachBatch)(gpuSortOrder)
-        case _: GpuHashAggregateExec =>
-          val gpuSortOrder = getOptimizedSortOrder(plan)
-          GpuSortExec(gpuSortOrder, false, plan, SortEachBatch)(gpuSortOrder)
+        case _: GpuDataWritingCommandExec =>
+          plan.withNewChildren(plan.children.map(c => insertHashOptimizeSorts(c, true)))
+        case hashJoin: GpuHashJoin =>
+          if (hasWriteParent) {
+            val gpuSortOrder = getOptimizedSortOrder(plan)
+            GpuSortExec(gpuSortOrder, false, hashJoin, SortEachBatch)(gpuSortOrder)
+          } else {
+            plan
+          }
+        case hashAgg: GpuHashAggregateExec =>
+          if (hasWriteParent) {
+            val gpuSortOrder = getOptimizedSortOrder(plan)
+            GpuSortExec(gpuSortOrder, false, hashAgg, SortEachBatch)(gpuSortOrder)
+          } else {
+            plan
+          }
```

Comment:
Since we treat the hash join and hash aggregate the same, we can simplify this a bit:
```suggestion
        case _: GpuHashJoin | _: GpuHashAggregateExec if hasWriteParent =>
            val gpuSortOrder = getOptimizedSortOrder(plan)
            GpuSortExec(gpuSortOrder, false, plan, SortEachBatch)(gpuSortOrder)
        case _: GpuHashJoin | _: GpuHashAggregateExec => plan
```